edb/pgsql/compiler/relgen.py

```diff
@@ -250,6 +250,10 @@ def _get_set_rvar(
         rvars = process_set_as_tuple_indirection(ir_set, stmt, ctx=ctx)
 
     elif isinstance(ir_set.expr, irast.FunctionCall):
+        if not ctx.session_mode and ir_set.expr.session_only:
+            raise RuntimeError(
```

Comment:
This check should be done in the EdgeQL translation phase, where we have the source context.  Also, `QueryError` should be raised, since this is a user error.

Reply:
OK, I've refactored the code to make this check at EdgeQL compile time.